### ISSUE TITLE
Dispose, destruct, reference counting of C# objects

### ISSUE THREAD

**Issue by simontom** [microsoft/ClearScript#525]
Hey guys,
is there any way how to recognize the object passed to JS got out of scope and can be disposed *automatically*?

Something like this:

C#
[ScriptMember]
ObjectToBeDisposed GetObject() => new ObjectToBeDisposed();

JS (for example for, or recursion, of just a scope of a function)
for (let i = 0; i < 42; i++) {
    const destructorObject = GetObject();
}

**Comment by simontom** [comment]
Well, I forgot, sorry....
What I have alredy tried is using a plain destructor `~ObjectToBeDisposed()` of the C# object. But, of course, it ain't a reliable way

**Comment by simontom** [comment]
Let's say
I have something like this

``` C#
class ObjectToBeDisposed: IDisposable
{
    void Dispose() => Console.WriteLine("Disposed");
}

class Service
{
    [ScriptMember("getObject")]
    ObjectToBeDisposed GetObject() => new ObjectToBeDisposed();
}
```

`Service`'s instance is passed to the CSEngine, so I could call `service.getobject()`

Now, I have something like this
``` Javascript
for (let i = 0; i < 42; i++) {
    const obj = service.getObject();
}
```

I understand:
- the .Net object gets wrapped into ScriptObject (_proxy object_ meant in ur post)
- when out of scope (here it is the for-body) it has no strong reference
- and after run out of scope it gets cleaned up when the GC kicks in

My question is if there is any way how we can recognize the loss of reference / getting out of scope, so, we would be able to call `Dispose` automatically

**Comment by ClearScriptLib** [comment]
Hi @simontom,

It sounds like you're looking for a hook that would allow you to call a managed object's `Dispose` method when the corresponding proxy is about to be collected by the script engine.

If that's correct, then, unfortunately, there's no such hook. Would it be possible for you to track your references manually? Your script code would probably have to adhere to some sort of acquire/release contract for these objects.

Good luck!

**Comment by simontom** [comment]
Yes, that's correct, something like a hook we could react to when the proxy is out of scope.

Do u have anything concrete up on your mind?
I see no way that the (even JS) object itself could be informed of not being used anymore (being out of scope, having no strong reference) as u wrote there is no such hook.

Well, it's OKay for us to add something like ˙JS ObjectToBeDisposed.close()˙. Was looking for a better, "no-code for user", solution.

Well, if you are done, told me everything already, let's close this issue.
